### PR TITLE
Provide an optional clip box instead of a mandatory bbox

### LINKED ISSUE #251: Consider allowing indirection to retrieve bounding box from different glyf glyph
In a conversation with @bungeman Ben points at a certain restriction when using a font as an outline font and COLRv1 font at the same time. I.e. displaying a COLRv1 font in an environment that does not understand COLRv1.

If the bounding box is always retrieved from the base glyph's entry in the `glyf` table, this fallback contour glyph would always need to be at least as wide as its COLRv1 rendering result (potentially widened by invisible points if needed).

Font makers would have more flexibility with the fallback `glyf` glyphs if we had a way to point to a dedicated bounding box glyph separately, other than always referring to the base glyph's `glyf` entry.

### PR DESCRIPTION
**Pull request by rsheeter**
Fixes #251. Again :)

### CODE REVIEW

File: OFF_AMD2_WD.md [hunk 1]
```diff
@@ -1140,18 +1140,20 @@ used for the color glyph. The advance width and height of glyphs referenced by
 PaintGlyph tables are not required to be the same as that of the base glyph and
 are ignored.
 
-For each color glyph, the bounding box for the color glyph is defined within
-the color glyph description (see 5.7.11.2.3). The bounding box data can be
-variable within a variable font.
+For each color glyph, a clip box may be associated (see 5.7.11.2.3). The clip
+box may vary in a variable font. If no clip is present it must be computed
+by traversing the graph of Paint that describes the contents of the glyph.
```
Comment (drott) [inline OFF_AMD2_WD.md:1145]:
```suggestion
box may vary in a variable font. If no clip is present if a bounding box is required by the implementation, it must be computed by traversing the graph of Paint tables that describes the contents of the glyph.
```

File: OFF_AMD2_WD.md [hunk 2]
```diff
@@ -1526,6 +1504,63 @@ are stacked above the top layer of this element.
 Offsets for paint tables not referenced by any PaintColrLayers table should not
 be included in the paintOffsets array.
 
+*ClipBox table:*
+
+The ClipBox table is used to provide a precomputed clip box for the given color glyph.
+Two formats are defined: format 0 for clip boxes without variation, and
+format 1 allowing for clip boxes that can vary in a variable font.
+
+Any content drawn outside the clip box should not render.
```
Comment (drott) [inline OFF_AMD2_WD.md:1513]:
```suggestion
Any content drawn outside the clip box shall not render.
```
shall, as in MUST  in RFCs?

File: OFF_AMD2_WD.md [hunk 3]
```diff
@@ -1526,6 +1504,63 @@ are stacked above the top layer of this element.
...
+| uint8 | format | Set to 1. |
+| FWORD | xMin | Minimum x of bounding box. For variation, use varIndexBase + 0. |
+| FWORD | yMin | Minimum y of bounding box. For variation, use varIndexBase + 1. |
+| FWORD | xMax | Maximum x of bounding box. For variation, use varIndexBase + 2. |
+| FWORD | yMax | Maximum y of bounding box. For variation, use varIndexBase + 3. |
+| uint32 | varIndexBase | Base index into DeltaSetIndexMap. |
+
+When computing a variable ClipBox compute the min/max coordinates using floating point
+values and then round to int values such that the clip box expands. That is, round xMin
+and yMin down and round xMax and yMax up.
```
Comment (drott) [inline OFF_AMD2_WD.md:1542]:
```suggestion
Note: When computing a variable ClipBox compute the min/max coordinates using floating point
values and then round to int values such that the clip box expands. That is, round xMin
and yMin down and round xMax and yMax up.
```

**Review by drott — CHANGES_REQUESTED** [review]
Mostly looks good to me. It'd be good to update the pseudocode as well:

> Allocate a bitmap for the glyph according to extents of base glyph contours for gid

could change to something like:

> Allocate a bitmap for the glyph according to the bounding box defined in a matching entry found from ClipList. If no entry is found, perform a glyph traversal to determine maximum bounds.

**Comment by rsheeter** [comment]
> update the pseudocode as well

Good catch, ty. Done.

**Comment by PeterConstable** [comment]
@googlebot I consent.

File: OFF_AMD2_WD.md [hunk 4]
```diff
@@ -1140,19 +1140,10 @@ used for the color glyph. The advance width and height of glyphs referenced by
...
-
-NOTE: If present, the clip box for the color glyph can be used to allocate a
-drawing surface without needing to traverse the graph of the color glyph
-definition.
-
 A valid color glyph definition shall define a bounded region—that is, it shall
 paint within a region for which a finite bounding box could be defined. The
 different paint formats have different boundedness characteristics:
 
-* If a clip box is provided the graph is inherantly bounded, no inspection
```
Comment (PeterConstable) [inline OFF_AMD2_WD.md]: The _graph_ isn't bounded. We're asserting that a clipbox makes the color glyph bounded.

Comment (PeterConstable) [reply OFF_AMD2_WD.md]: Also, moved out of that list since the list is about paint tables.

File: OFF_AMD2_WD.md [hunk 5]
```diff
@@ -1168,19 +1159,31 @@ sub-graph is bounded.
...
+definition.
+
+If a clip box is provided for a color glyph, the color glyph is bounded, and no
+inspection of the Paint graph is required to determine boundedness. If no clip
+box is defined for a color glyph, however, applications shall confirm that the
+color glyph definition is bounded, and shall not render the color glyph if the
+defining graph is not bounded.
 
 To ensure that rendering implementations do not clip any part of a color glyph,
-the bounding box of the base glyph needs to be large enough to encompass the
```
Comment (PeterConstable) [inline OFF_AMD2_WD.md:1174]: No longer need to talk about the bbox of the base glyph.

**Comment by PeterConstable** [comment]
Added a commit with various editorial changes. In 5.7.11.2.3, I reversed the order in which the list/record/table structs are presented to match what is done elsewhere. Also changed the clip record range field names to match what is done elsewhere.

**Comment by PeterConstable** [comment]
@googlebot I consent.

**Comment by PeterConstable** [comment]
@googlebot I consent.

**Comment by PeterConstable** [comment]
@googlebot I consent.

File: OFF_AMD2_WD.md (hunk 5 above)
Comment (rsheeter) [reply OFF_AMD2_WD.md:1176]: Good catch.

**Review by PeterConstable — APPROVED** [review]

File: OFF_AMD2_WD.md (hunk 4 above)
Comment (rsheeter) [reply OFF_AMD2_WD.md]: I believe you fixed this (ty) so resolving

**Pull request merged**